Update deploy script with getopts and logging

jasoncpatton opened the pull request:
1. Use getopts to optionally take central manager, data	source name, and data source directory
2. Fallback to reading user input for these
3. Output progress to stdout
4. Log bash's xtrace and command stdout/stderr to /tmp/pid.log
5. Fix permissions on data source directory to other-readable

When merged, resolves #2 and #3

jasoncpatton committed:
Update deploy script with getopts and logging

1. Use getopts to optionally take central manager, data	source name, and data source directory
2. Fallback to reading user input for these
3. Output progress to stdout
4. Log bash's xtrace and command stdout/stderr to /tmp/pid.log
5. Fix permissions on data source directory to other-readable

bbockelm commented:
@matyasselmeci can you review?

jasoncpatton committed:
Fix args

jasoncpatton committed:
Give apt a second to complete

jasoncpatton committed:
Squelch grep of apt sources list

jasoncpatton committed:
Break final status message

matyasselmeci reviewed (commented):
Mostly fine, though I recommend adding some sanitization for `$CENTRAL_MANAGER` and `$DATA_SOURCE_NAME` -- this would work (though the hostname check is lazy):
```
[[ $CENTRAL_MANAGER =~ [a-z0-9][a-z0-9.-]* ]] || \
    fail "The central manager name should be a valid hostname"
[[ $DATA_SOURCE_NAME =~ [A-Za-z0-9_]+ ]] || \
    fail "The data source name should only contain alphanumeric characters and underscores"
```
More importantly, you **really** ought to add some safety to `$DATA_SOURCE_DIRECTORY`: you never know when some user is going to accidentally use `/` and completely break their whole system. Do `DATA_SOURCE_DIRECTORY=$(readlink -f "$DATA_SOURCE_DIRECTORY")` to canonicalize the path (convert to an absolute path, resolve symlinks, get rid of extra slashes) and check it against a blacklist that contains the standard system paths (`"/", "/bin", "/etc", ...`),

jasoncpatton committed:
Run sanity checks on input variables

Following @matyasselmeci's advice, the script now checks:
1. For a valid CM hostname (not that it exists, but it could exist)
2. For a valid data source name
3. For a valid data source directory, by checking for...
  a. Existence (this was already checked)
  b. Absolute path
  c. Real path (e.g. follow symlinks)
  d. Not a system path

jasoncpatton commented:
> Mostly fine, though I recommend adding some sanitization for `$CENTRAL_MANAGER` and `$DATA_SOURCE_NAME` -- this would work (though the hostname check is lazy):
> 
> ```
> [[ $CENTRAL_MANAGER =~ [a-z0-9][a-z0-9.-]* ]] || \
>     fail "The central manager name should be a valid hostname"
> [[ $DATA_SOURCE_NAME =~ [A-Za-z0-9_]+ ]] || \
>     fail "The data source name should only contain alphanumeric characters and underscores"
> ```
> 
> More importantly, you **really** ought to add some safety to `$DATA_SOURCE_DIRECTORY`: you never know when some user is going to accidentally use `/` and completely break their whole system. Do `DATA_SOURCE_DIRECTORY=$(readlink -f "$DATA_SOURCE_DIRECTORY")` to canonicalize the path (convert to an absolute path, resolve symlinks, get rid of extra slashes) and check it against a blacklist that contains the standard system paths (`"/", "/bin", "/etc", ...`),

Thanks Mat! I think I've managed to incorporate these checks, although they may not be completely userproof. Still much better than before.

matyasselmeci commented:
Looks pretty good. If you merge it, I can test the new version.

bbockelm commented:
@jasoncpatton - please go ahead and merge without me.  Need to attend to a few other projects right now.
